internal/state/change_processor_test.go

```diff
@@ -243,1023 +221,132 @@ var _ = Describe("ChangeProcessor", func() {
-								PathRules: []dataplane.PathRule{
-									{
-										Path:     "/",
-										PathType: dataplane.PathTypePrefix,
-										MatchRules: []dataplane.MatchRule{
-											{
-												MatchIdx:     0,
-												RuleIdx:      0,
-												BackendGroup: hr1Group,
-												Source:       hr1,
-											},
-										},
-									},
-								},
-							},
-						},
-						SSLServers: []dataplane.VirtualServer{
-							{
-								IsDefault: true,
-							},
-							{
-								Hostname: "foo.example.com",
-								SSL:      &dataplane.SSL{CertificatePath: certificatePath},
-								PathRules: []dataplane.PathRule{
-									{
-										Path:     "/",
-										PathType: dataplane.PathTypePrefix,
-										MatchRules: []dataplane.MatchRule{
-											{
-												MatchIdx:     0,
-												RuleIdx:      0,
-												BackendGroup: hr1Group,
-												Source:       hr1,
-											},
-										},
-									},
-								},
-							},
-							{
-								Hostname: "~^",
-								SSL:      &dataplane.SSL{CertificatePath: certificatePath},
-							},
-						},
-						BackendGroups: []dataplane.BackendGroup{hr1Group},
-					}
-
-					expectedStatuses := state.Statuses{
-						GatewayClassStatus: &state.GatewayClassStatus{
-							ObservedGeneration: gc.Generation,
-							Conditions:         conditions.NewDefaultGatewayClassConditions(),
-						},
-						GatewayStatuses: state.GatewayStatuses{
-							{Namespace: "test", Name: "gateway-1"}: {
-								Conditions:         conditions.NewDefaultGatewayConditions(),
-								ObservedGeneration: gw1.Generation,
-								ListenerStatuses: map[string]state.ListenerStatus{
-									"listener-80-1": {
-										AttachedRoutes: 1,
-										Conditions:     conditions.NewDefaultListenerConditions(),
-									},
-									"listener-443-1": {
-										AttachedRoutes: 1,
-										Conditions:     conditions.NewDefaultListenerConditions(),
-									},
-								},
-							},
-						},
-						HTTPRouteStatuses: state.HTTPRouteStatuses{
-							{Namespace: "test", Name: "hr-1"}: {
-								ObservedGeneration: hr1.Generation,
-								ParentStatuses: []state.ParentStatus{
-									{
-										GatewayNsName: client.ObjectKeyFromObject(gw1),
-										SectionName:   helpers.GetPointer[v1beta1.SectionName]("listener-80-1"),
-										Conditions:    conditions.NewDefaultRouteConditions(),
-									},
-									{
-										GatewayNsName: client.ObjectKeyFromObject(gw1),
-										SectionName:   helpers.GetPointer[v1beta1.SectionName]("listener-443-1"),
-										Conditions:    conditions.NewDefaultRouteConditions(),
-									},
-								},
-							},
-						},
-					}
-
-					changed, conf, statuses := processor.Process(context.TODO())
+					changed, graphCfg := processor.Process()
 					Expect(changed).To(BeTrue())
-					Expect(helpers.Diff(expectedConf, conf)).To(BeEmpty())
-					assertStatuses(expectedStatuses, statuses)
+					Expect(graphCfg).ToNot(BeNil())
```

Comment:
why not compare against an initialized Graph struct?

Because we're doing this shortcut, I wonder if it makes sense to go once step further and separate state from processing? So that we can separate tests that test state-change-checking logic from building the graph.

This means, in event handler, we can have something like below:

```go
func (h *EventHandlerImpl) HandleEventBatch(ctx context.Context, batch EventBatch) {
	for _, event := range batch {
		switch e := event.(type) {
		case *UpsertEvent:
			h.state.Upsert(e)
		case *DeleteEvent:
			h.state.Delete(e)
		default:
			panic(fmt.Errorf("unknown event type %T", e))
		}
	}

	if !state.Changed() {
		h.cfg.Logger.Info("Handling events didn't result into NGINX configuration changes")
		return
	}

	// graphClusterState includes maps that back h.state
	graphCfg := graph.BuildGraph(graphClusterState, ...)

	var nginxReloadRes status.NginxReloadResult
	err := h.updateNginx(ctx, dataplane.BuildConfiguration(ctx, graphCfg, h.cfg.ServiceResolver))
	if err != nil {
		h.cfg.Logger.Error(err, "Failed to update NGINX configuration")
		nginxReloadRes.Error = err
	} else {
		h.cfg.Logger.Info("NGINX configuration was successfully updated")
	}

	h.cfg.StatusUpdater.Update(ctx, status.BuildStatuses(graphCfg, nginxReloadRes))
}
```

Reply:
ChangeProcessor doesn't build the graph, that's for the graph package. Which should be tested in that package, not here. Just becomes redundant.

I think I commented this on an earlier thread, but I didn't want to expose the internal state held by the change processor to the event handler. I like the separation of concerns, being that the event handler sees an update, and then delegates to the component that actually holds the state information to build the configuration.

Reply:
Did we lose any test coverage by removing these tests, or were they completely redundant?

Looking at the code in Process, the graph is built off of the cluster state, which is maintained by the ChangeProcessor. Without testing the contents of the graph that's returned, can we be sure that the cluster state is correct?

Reply:
My concern is that we are no longer testing the accuracy of the entire cluster state; we are only testing whether the state has changed and if the graph is built. The graph tests verify that we build the correct graph given a cluster state, but we aren't testing that the ChangeProcessor builds the correct cluster state.

Reply:
Maybe we need a whole new set of tests for the store itself and its contents? We have tests verifying that Upserts/Deletes result in changes, so we have that covered at least.

Reply:
I'm finding even more duplicate tests in change processor though.

Reply:
I'm finding even more duplicate tests in the change processor though.

Reply:
Let me work on updating the change processor tests to verify the graph contents.

Reply:
@kate-osborn Added back the tests from before, but simplified the struct verification. Still checking if the expected fields are set, just not checking the entire graph config since that became a bit unwieldy.

Reply:
Looks better. I'm still worried we lost some test coverage here, but I appreciate these tests' simplicity. Is it worth adding a single test that processes a bunch of upserts and deletes and then asserts on the entire structure of the graph? 

@pleshakov curious what you think.

Reply:
> I still feel like that's more of a graph test. If we've verified that the changeprocessor upserts/deletes populate the resources we want, do we care about the specifics of all the fields?

I think if we don't test what the state contains (whether through building a graph or by looking into the maps), we can't be sure it has been updated correctly. That's because the state does a number of things - tracks relationships, performs validation.

For example, in a test we can upsert a resource and then confirm that the graph has not been generated (state hasn't changed). However, that could happen not only if the resource has not been changed, but also if the resource is invalid (fails webhook validation). By not looking into the state or the graph, we can't be sure.

However, let's acknowledge that we had shortcuts before this PR too:
```
		// Note: in these tests, we deliberately don't fully inspect the returned configuration and statuses
		// -- this is done in 'Normal cases of processing changes'
```

Ultimately, I think we should extensively tests the output of the the thing under the test. If the current output is the graph, we shall test that it is correct. If that gets too complicated, perhaps it makes sense to have separate tests for the state part component?

Reply:
I've added the checks for the graph structure.

I'm still not convinced that the bulk of these tests have to be in the change processor since we can provide the state input to the graph tests directly, while also testing the change processor to ensure it determines state changes properly, but I'll concede :).